Installation issue: llvm@14.0.0, 'hwloc.h' file not found at build time despite it being found during configure

### Steps to reproduce the issue

```
$ spack install -j4 llvm@14.0.0+libcxx+omp_as_runtime+omp_debug+omp_tsan%gcc@11.2.0
```

Concertized spec:
```
sajid@LAPTOP-CDJT2P3R ~> spack spec -I llvm@14.0.0+libcxx+omp_as_runtime+omp_debug+omp_tsan%gcc@11.2.0
Input spec
--------------------------------
 -   llvm@14.0.0%gcc@11.2.0+libcxx+omp_as_runtime+omp_debug+omp_tsan

Concretized
--------------------------------
 -   llvm@14.0.0%gcc@11.2.0+clang~code_signing+compiler-rt~cuda~flang+gold+internal_unwind~ipo+libcxx~link_llvm_dylib+lld+lldb~llvm_dylib~mlir+omp_as_runtime+omp_debug+omp_tsan+polly~python~shared_libs~split_dwarf~z3 build_type=Release targets=none version_suffix=none arch=linux-ubuntu20.04-zen2
[+]      ^binutils@2.38%gcc@11.2.0~gas+gold~headers~interwork+ld~libiberty~lto+nls+plugins libs=shared,static arch=linux-ubuntu20.04-zen2
[+]          ^diffutils@3.8%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]              ^libiconv@1.16%gcc@11.2.0 libs=shared,static arch=linux-ubuntu20.04-zen2
[+]          ^gettext@0.21%gcc@11.2.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu20.04-zen2
[+]              ^bzip2@1.0.8%gcc@11.2.0~debug~pic+shared arch=linux-ubuntu20.04-zen2
[+]              ^libxml2@2.9.12%gcc@11.2.0~python arch=linux-ubuntu20.04-zen2
[+]                  ^pkgconf@1.8.0%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]                  ^xz@5.2.5%gcc@11.2.0~pic libs=shared,static arch=linux-ubuntu20.04-zen2
[+]                  ^zlib@1.2.11%gcc@11.2.0+optimize+pic+shared arch=linux-ubuntu20.04-zen2
[+]              ^ncurses@6.2.20200212%gcc@11.2.0~symlinks+termlib abi=6 arch=linux-ubuntu20.04-zen2
[+]              ^tar@1.34%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]      ^cmake@3.22.3%gcc@11.2.0~doc+ncurses+openssl+ownlibs~qt build_type=Release arch=linux-ubuntu20.04-zen2
[+]          ^openssl@1.1.1n%gcc@11.2.0~docs~shared certs=system arch=linux-ubuntu20.04-zen2
[+]              ^perl@5.34.0%gcc@11.2.0+cpanm+shared+threads arch=linux-ubuntu20.04-zen2
[+]                  ^berkeley-db@18.1.40%gcc@11.2.0+cxx~docs+stl patches=b231fcc arch=linux-ubuntu20.04-zen2
[+]                  ^gdbm@1.19%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]                      ^readline@8.1%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]      ^hwloc@2.7.0%gcc@11.2.0~cairo~cuda~gl~libudev+libxml2~netloc~nvml~opencl+pci~rocm+shared arch=linux-ubuntu20.04-zen2
[+]          ^libpciaccess@0.16%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]              ^libtool@2.4.6%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]                  ^m4@1.4.19%gcc@11.2.0+sigsegv patches=9dc5fbd,bfdffa7 arch=linux-ubuntu20.04-zen2
[+]                      ^libsigsegv@2.13%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]              ^util-macros@1.19.3%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]      ^libedit@3.1-20210216%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]      ^ninja@1.10.2%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]          ^python@3.9.10%gcc@11.2.0+bz2+ctypes+dbm~debug+ensurepip+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93,4c24573,f2fd060 arch=linux-ubuntu20.04-zen2
[+]              ^expat@2.4.6%gcc@11.2.0+libbsd arch=linux-ubuntu20.04-zen2
[+]                  ^libbsd@0.11.5%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]                      ^libmd@1.0.4%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]              ^libffi@3.4.2%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]              ^sqlite@3.37.2%gcc@11.2.0+column_metadata+dynamic_extensions+fts~functions+rtree arch=linux-ubuntu20.04-zen2
[+]              ^util-linux-uuid@2.37.4%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]      ^perl-data-dumper@2.173%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]      ^swig@4.0.2%gcc@11.2.0 arch=linux-ubuntu20.04-zen2
[+]          ^pcre@8.45%gcc@11.2.0~jit+multibyte+utf arch=linux-ubuntu20.04-zen2

sajid@LAPTOP-CDJT2P3R ~>
```



### Error message

Relevant error message: 

```
  >> 26468    FAILED: openmp/libompd/src/CMakeFiles/ompd.dir/omp-icv.cpp.o
     26469    /tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-build-lmdeeji/./bin/clang++ --target=x86_64-unknown-linux-gnu -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Dom
              pd_EXPORTS -I/tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-src/openmp/libompd/src -I/tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-build-lmdeej
              i/runtimes/runtimes-bins/openmp/runtime/src -I/tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-src/openmp/runtime/src -std=c++11 -fPIC -fno-semantic-interposition -fvisibility-inlines-h
              idden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wn
              on-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wcast-qual -Wformat-pedantic -Wimplicit-
              fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -std=c++14 -std=c++11 -O3 -DNDEBUG -fPIC -MD -MT openmp/libompd/src/CMakeFiles/ompd.dir/omp-icv.cpp.o -MF openmp/libompd/src/CMakeFiles/ompd.dir/omp-icv.cpp.o.d -o openmp
              /libompd/src/CMakeFiles/ompd.dir/omp-icv.cpp.o -c /tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-src/openmp/libompd/src/omp-icv.cpp
     26470    In file included from /tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-src/openmp/libompd/src/omp-icv.cpp:21:
  >> 26471    /tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-src/openmp/runtime/src/kmp.h:96:10: fatal error: 'hwloc.h' file not found
     26472    #include "hwloc.h"
     26473             ^~~~~~~~~
     26474    1 error generated.
```

This occurs, despite the fact that spack passes `'-DLIBOMP_USE_HWLOC:BOOL=ON' '-DLIBOMP_HWLOC_INSTALL_DIR:STRING=/home/sajid/packages/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-11.2.0/hwloc-2.7.0-5ttjykgyd7bxnhegnydeesc53ufojjmb' ` and the fact that `hwloc` is indeed found during configuration (for `runtimes-bins`) : 
```
-- Looking for hwloc.h
-- Looking for hwloc.h - found
-- Looking for hwloc_topology_init in /home/sajid/packages/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-11.2.0/hwloc-2.7.0-5ttjykgyd7bxnhegnydeesc53ufojjmb/lib/libhwloc.so
-- Looking for hwloc_topology_init in /home/sajid/packages/spack/opt/spack/linux-ubuntu20.04-zen2/gcc-11.2.0/hwloc-2.7.0-5ttjykgyd7bxnhegnydeesc53ufojjmb/lib/libhwloc.so - found
```

For reference, the generated build recipe (`spack-build-lmdeeji/runtimes/runtimes-bins/build.ninja`) for the failing file is: 
```
build openmp/libompd/src/CMakeFiles/ompd.dir/omp-icv.cpp.o: CXX_COMPILER__ompd_Release /tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-src/openmp/libompd/src/omp-icv.cpp || cmake_object_order_depends_target_ompd
  DEFINES = -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Dompd_EXPORTS
  DEP_FILE = openmp/libompd/src/CMakeFiles/ompd.dir/omp-icv.cpp.o.d
  FLAGS = -std=c++11 -fPIC -fno-semantic-interposition -fvisibility-inlines-hidden -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wnon-virtual-dtor -Wdelete-non-virtual-dtor -Wsuggest-override -Wno-comment -Wstring-conversion -Wmisleading-indentation -fdiagnostics-color -ffunction-sections -fdata-sections -Wall -Wcast-qual -Wformat-pedantic -Wimplicit-fallthrough -Wsign-compare -Wno-extra -Wno-pedantic -std=c++14 -std=c++11 -O3 -DNDEBUG -fPIC
  INCLUDES = -I/tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-src/openmp/libompd/src -I/tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-build-lmdeeji/runtimes/runtimes-bins/openmp/runtime/src -I/tmp/sajid/spack-stage/spack-stage-llvm-14.0.0-lmdeejip2qzq4j4ntm7tbgx6pqua2z26/spack-src/openmp/runtime/src
  OBJECT_DIR = openmp/libompd/src/CMakeFiles/ompd.dir
  OBJECT_FILE_DIR = openmp/libompd/src/CMakeFiles/ompd.dir
  TARGET_COMPILE_PDB = openmp/libompd/src/CMakeFiles/ompd.dir/
  TARGET_PDB = openmp/libompd/src/libompd.pdb
```

### Information on your system

```
sajid@LAPTOP-CDJT2P3R ~> spack debug report
* **Spack:** 0.18.0.dev0 (1dbab78db18629ba3c606309814898fe9d86e7a8)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-zen2
* **Concretizer:** clingo
sajid@LAPTOP-CDJT2P3R ~>
```

### Additional information

[spack-build-01-cmake-out.txt](https://github.com/spack/spack/files/8364798/spack-build-01-cmake-out.txt)
[spack-build-02-build-out.txt](https://github.com/spack/spack/files/8364796/spack-build-02-build-out.txt)

Tagging @trws & @haampie

Though the hwloc arguments were fixed in https://github.com/spack/spack/pull/16104, this is a new issue as LLVM14 includes the `ompd` library (which includes the `omp-icv.cpp` source file that fails to build).

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/llvm/llvm14-hwloc-ompd.patch
+++ b/var/spack/repos/builtin/packages/llvm/llvm14-hwloc-ompd.patch
@@ -1,0 +1,13 @@
+--- a/openmp/libompd/src/CMakeLists.txt
++++ b/openmp/libompd/src/CMakeLists.txt
+@@ -44,6 +44,10 @@
+         ${LIBOMP_SRC_DIR}
+ )
+
++if(${LIBOMP_USE_HWLOC})
++  include_directories(${LIBOMP_HWLOC_INSTALL_DIR}/include)
++endif()
++
+ INSTALL( TARGETS ompd
+         LIBRARY DESTINATION ${OPENMP_INSTALL_LIBDIR}
+         ARCHIVE DESTINATION ${OPENMP_INSTALL_LIBDIR}

--- a/var/spack/repos/builtin/packages/llvm/package.py
+++ b/var/spack/repos/builtin/packages/llvm/package.py
@@ -358,6 +358,9 @@ class Llvm(CMakePackage, CudaPackage):
     # avoid build failed with Fujitsu compiler
     patch('llvm13-fujitsu.patch', when='@13 %fj')
 
+    # patch for missing hwloc.h include for libompd
+    patch('llvm14-hwloc-ompd.patch', when='@14')
+
     # The functions and attributes below implement external package
     # detection for LLVM. See:
     #

